File: reservoir/submissions/accepted/reservoir_dsu.cpp
#include <iostream>
#include <set>
#include <vector>
#include <map>
#include <algorithm>

using namespace std;
typedef vector<int> vi;
typedef vector<vi> v2i;
typedef pair<int, int> pi2;
typedef vector<pi2> vp;

int islands;

struct DSU {
	vi p, r; // Parents and ranks/depths of elements

	DSU(int n) {
		for (int i = 0; i < n; i++) p.push_back(i);
		r.resize(n);
	}
	
	int get(int a) {
		return p[a] = a == p[a] ? a : get(p[a]);
	}
	void uni(int a, int b) {
		a = get(a), b = get(b);
		if (a == b) return;
		if (r[b] < r[a]) swap(a, b);
		if (r[b] == r[a]) r[b]++;
		p[a] = b;
		islands--;
	}
};

const int DIRS[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

/**
 * Return the maximum number of islands.
 *
 * N: number of rows
 * M: number of columns
 * G: grid of heights
 */
int solve(int N, int M, vector<vector<int>> &G) {
    // G' is G with borders (-1 will make the cell always submerged)
    v2i Gp(N + 2, vi(M + 2, -1));
    for (int i = 0; i < N; i++)
        for (int j = 0; j < M; j++)
            Gp[i + 1][j + 1] = G[i][j];

    map<int, vp> m; // Maps each height to the set of cells at that height
    for (int i = 1; i <= N; i++)
        for (int j = 1; j <= M; j++)
            // Using negative height to help us iterate in reverse later 
            m[-Gp[i][j]].emplace_back(i, j);
    
    // DSU in reverse
    DSU dsu((N + 2) * (M + 2));
    int mx_islands = 0;
    islands = 0;
    for (pair<int, vp> p: m) {
        int h = -p.first;
        vp cells = p.second;
        islands += cells.size();
        for (pi2 cell: cells) {
            int i = cell.first, j = cell.second;
            for (auto d: DIRS) {
                int ip = i + d[0], jp = j + d[1];
                if (Gp[ip][jp] >= h)
                    dsu.uni(i * (M + 2) + j, ip * (M + 2) + jp);
            }
        }
        mx_islands = max(mx_islands, islands);
    }

    return mx_islands;
}

int main() {
    int T;
    cin >> T;
    for (int t = 0; t < T; t++) {
        int N, M;
        cin >> N >> M;
        vector<vector<int>> G(N, vector<int>(M));
        for (int i = 0; i < N; i++)
            for (int j = 0; j < M; j++)
                cin >> G[i][j];
        cout << solve(N, M, G) << '\n';
    }
}

File: reservoir/submissions/accepted/reservoir_dsu.py
from typing import List, Tuple, Dict
from collections import defaultdict

class DSU:
    def __init__(self, n: int):
        self.p = list(range(n))
        self.r = [0] * n

    def get(self, a: int) -> int:
        if a != self.p[a]:
            self.p[a] = self.get(self.p[a])
        return self.p[a]

    def uni(self, a: int, b: int):
        a, b = self.get(a), self.get(b)
        if a == b:
            return
        if self.r[b] < self.r[a]:
            a, b = b, a
        if self.r[b] == self.r[a]:
            self.r[b] += 1
        self.p[a] = b
        global islands
        islands -= 1

DIRS = [(-1, 0), (1, 0), (0, -1), (0, 1)]

def solve(N: int, M: int, G: List[List[int]]) -> int:
    global islands
    # Extend grid with borders (-1 for always-submerged border cells)
    Gp = [[-1] * (M + 2) for _ in range(N + 2)]
    for i in range(N):
        for j in range(M):
            Gp[i + 1][j + 1] = G[i][j]

    # Mapping height to list of cells at that height
    m = {}
    for i in range(1, N + 1):
        for j in range(1, M + 1):
            key = Gp[i][j]
            if key not in m:
                m[key] = []
            m[key].append((i, j))

    dsu = DSU((N + 2) * (M + 2))
    mx_islands = 0
    islands = 0
    for h, cells in sorted(m.items())[::-1]:
        islands += len(cells)
        for i, j in cells:
            for di, dj in DIRS:
                ip, jp = i + di, j + dj
                if Gp[ip][jp] >= h:
                    dsu.uni(i * (M + 2) + j, ip * (M + 2) + jp)
        mx_islands = max(mx_islands, islands)

    return mx_islands

def main():
    T = int(input())
    for _ in range(T):
        N, M = map(int, input().split())
        G = [list(map(int, input().split())) for _ in range(N)]
        print(solve(N, M, G))

if __name__ == "__main__":
    main()
